### PR TITLE
make it possible to stream public timelines without authorization

### PR DESCRIPTION
**Pull request by nullkal**
Public timelines can be fetched via `/api/v1/timelines/public` without authorization, but streaming them needs authorization token now. This PR make it consistent between normal fetching APIs and streaming APIs.

### CODE REVIEW

File: streaming/index.js [hunk 1]
```diff
@@ -227,6 +227,19 @@ const startWorker = (workerId) => {
   };
 
   const wsVerifyClient = (info, cb) => {
+    const PUBLIC_STREAMS = [
```
Comment (abcang) [inline streaming/index.js]: I think that it is better to define PUBLIC_STREAMS and PUBLIC_ENDPOINTS outside the function.

**Comment by abcang** [comment]
If authentication is skipped, since `req.accountId` does not have a value, I think you need to change some methods such as streamFrom. Especially streamFrom will filter on accounts if `needsFiltering` is true, so we need to fix.

File: streaming/index.js [hunk 2]
```diff
@@ -208,42 +208,65 @@ const startWorker = (workerId) => {
...
+  ];
+
   const authenticationMiddleware = (req, res, next) => {
     if (req.method === 'OPTIONS') {
       next();
       return;
     }
 
-    accountFromRequest(req, next);
+    const auth_required = PUBLIC_ENDPOINTS.some(endpoint => endpoint === req.path);
```
Comment (abcang) [inline streaming/index.js]: Maybe `!PUBLIC_ENDPOINTS.some(endpoint => endpoint === req.path)`. Also, it is better to change the variable name to camel case.

File: streaming/index.js [hunk 3]
```diff
@@ -313,26 +337,30 @@ const startWorker = (workerId) => {
             return;
           }
 
-          const queries = [
-            client.query(`SELECT 1 FROM blocks WHERE (account_id = $1 AND target_account_id IN (${placeholders(targetAccountIds, 2)})) OR (account_id = $2 AND target_account_id = $1) UNION SELECT 1 FROM mutes WHERE account_id = $1 AND target_account_id IN (${placeholders(targetAccountIds, 2)})`, [req.accountId, unpackedPayload.account.id].concat(targetAccountIds)),
-          ];
+          if (req.accountId != null) {
```
Comment (abcang) [inline streaming/index.js]: I think that simply does not have any problem. `accountId` can never be 0.

```
if (!req.accountId) {
```

And you should not need to change `.eslintrc.yml`.

**Review by abcang — CHANGES_REQUESTED** [review]

**Review by abcang — APPROVED** [review]

File: streaming/index.js [hunk 4]
```diff
@@ -383,16 +411,18 @@ const startWorker = (workerId) => {
 
   // Setup stream end for WebSockets
   const streamWsEnd = (req, ws, closeHandler = false) => (id, listener) => {
+    const accountId = req.accountId || '(anonymous user)';
```
Comment (Gargron) [inline streaming/index.js]: Can we use IP instead?

File: streaming/index.js [hunk 5]
```diff
@@ -275,8 +297,10 @@ const startWorker = (workerId) => {
   };
 
   const streamFrom = (id, req, output, attachCloseHandler, needsFiltering = false, notificationOnly = false) => {
+    const accountId = req.accountId || '(anonymous user)';
```
Comment (Gargron) [inline streaming/index.js]: Can we use IP instead?

**Comment by Gargron** [comment]
I think conceptually I am okay with this. Originally I was against it because keeping open connections is a burden and allowing anonymous connections means you would have to block by IP instead of simply revoking access tokens.

However if this becomes a thing than the /about page could get real-time updates instead of poll intervals.

**Review by Gargron — APPROVED** [review]

**Pull request merged**